docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,562 @@
+   unless the user tells datamon these may be safely dropped.
+7. Details about conflicts and checkpoints are retained as part of the bundled dataset, in some "hidden folders"
+8. When users explicitly mention their intent to carry out _incremental uploads_, conflicts are renamed as checkpoints,
+   thus establishing a distinction between "good" and "bad" conflicts.
+9. When users explicitly mention their intent to ignore conflicts, conflicts are discarded and older versions of files
+   simply dropped at commit time.
+
+> **NOTE**: we don't want to upload blobs without keeping a track of the reference hashes
+> (unrecoverable blobs), unless explicitly being told to.
+
+
+Design:
+
+* Latest version wins: incremental uploads overwriting the same file with several versions always ends up with a
+  safe deconfliction.
+* References to all conflicting files, save the winner, will be stored in a special path on the dataset available
+  with the bundle, and may be readily downloaded:
+  ```
+  /.conflicts/{split-id}/{path}
+  ```
+  or:
+  ```
+  /.checkpoints/{split-id}/{path}
+  ```
+* Conflicts and checkpoints in these special locations won't be uploaded if the download is reused to create a new
+  bundle (i.e. `/.conflicts`, `/.checkpoints` are ignored by uploads just as are the metadata in `.datamon`).
+
+> **NOTE**: conflict resolution is based on the _upload time_ of the file (not modification time on the dataset), and
+> is resolved locally by the host running the split client.
+>
+> Therefore, hosts running splits should be synchronized.
+
+
+### Handling failures, retries, and cancellations
+
+Premises:
+
+1. Since the commit action is user-driven, there must be no limit set on the delay between the completion of the splits
+   and the final commit.
+2. Some `add` operations might fail: the client has the opportunity to either _retry_ any individual split or _cancel_
+   the whole diamond operation.
+3. This is different from the `bundle upload` operation, which only succeeds or fails as a whole.
+   Since we now leave it to the user to decide if and when to commit, there is a newly available choice to retry on
+   individual failed splits.
+4. Blobs put on storage resulting from split uploads cannot be relinquished. We accept the possibility of some wasted
+   storage as a result.
+5. Failure to carry out a commit is a special case, see [below](#failed-commit).
+
+6. **Nice to have**. Explicit cancellation may be skipped and does not affect concurrent diamonds or diamonds initiated
+   in the future. Explicit cancellation is only expected from well-behaved users and helps maintain orderly metadata.
+   This is a consequence of the blob deduplication strategy (the same blob might be claimed by another file, so we can't
+   mark it as something to be deleted).
+
+
+Design:
+
+* Diamonds abide by the following state-transition model:
+  1. `initialized`
+  2. `committing`
+  3. `done`
```

Comment:
Instead of tracking this as fields in a single yaml file which will require a Read-Modify-Write cycle which under concurrent load is hard to get right. Use separate files and use the put if not present primitive. Also we only need init and done.

Reply:
I don't see the need for a 2 phase commit verbiage.

Reply:
_verbiage_ is perhaps not carrying the same meaning. Certainly not when compared with some _actual_ verbiage that clutters the other reviews on this series of PR.
Best to discuss that part separately.

Reply:
moved back to fewer states for diamond: initialized (aka running), done, canceled

and for splits: running, done

Reply:
Even though the key name now holds the state semantics, I am keeping the state encoded in the metadata for easier retrieval (it is a bit of a pain to query the state of things when all is encoded as keys)